File: src/utils/scheduler.ts
/**
 * @license
 * SPDX-License-Identifier: Apache-2.0
 */

import { Team, Match } from "../types";

/**
 * Thuật toán xếp lịch thi đấu vòng tròn 1 lượt (Circle Method / Round-robin)
 * Áp dụng cơ chế xoay vòng. Nếu số đội lẻ, thêm đội "BYE" (Nghỉ).
 */
export function generateRoundRobin(teams: Team[], groupName?: string): Match[] {
  if (teams.length < 2) return [];

  // Tạo bản sao danh sách đội
  let list = [...teams];
  
  // Nếu lẻ, thêm đội ảo "Nghỉ" (BYE)
  const isOdd = list.length % 2 !== 0;
  if (isOdd) {
    list.push({ id: "BYE", name: "Nghỉ", group: groupName });
  }

  const numTeams = list.length;
  const numRounds = numTeams - 1;
  const matches: Match[] = [];

  // Để tạo thứ tự vòng đấu mượt mà và tránh trùng lặp
  for (let round = 1; round <= numRounds; round++) {
    for (let i = 0; i < numTeams / 2; i++) {
      const homeIdx = i;
      const awayIdx = numTeams - 1 - i;
      const home = list[homeIdx];
      const away = list[awayIdx];

      // Đảo vị trí sân nhà/sân khách xen kẽ để công bằng
      const isHome = round % 2 === 1 ? i % 2 === 0 : i % 2 === 1;
      const first = isHome ? home : away;
      const second = isHome ? away : home;

      if (first.id !== "BYE" && second.id !== "BYE") {
        matches.push({
          id: `${groupName || "RR"}-r${round}-${first.id}-${second.id}`,
          round,
          roundName: `Vòng ${round}`,
          group: groupName,
          homeTeamId: first.id,
          awayTeamId: second.id,
          played: false,
          homeYellowCards: 0,
          homeRedCards: 0,
          awayYellowCards: 0,
          awayRedCards: 0,
        });
      } else {
        // Trận đấu nghỉ: Đội còn lại được nghỉ ở vòng này
        const activeTeam = first.id === "BYE" ? second : first;
        matches.push({
          id: `${groupName || "RR"}-r${round}-${activeTeam.id}-bye`,
          round,
          roundName: `Vòng ${round}`,
          group: groupName,
          homeTeamId: activeTeam.id,
          awayTeamId: "BYE",
          played: true, // Coi như đã xong
          homeScore: 0,
          awayScore: 0,
          homeYellowCards: 0,
          homeRedCards: 0,
          awayYellowCards: 0,
          awayRedCards: 0,
          isBye: true,
        });
      }
    }
    // Xoay vòng tròn: giữ phần tử đầu tiên cố định, xoay các phần tử còn lại
    list = [list[0], list[list.length - 1], ...list.slice(1, list.length - 1)];
  }

  // Sắp xếp các trận đấu theo thứ tự vòng đấu cho gọn
  return matches.sort((a, b) => a.round - b.round);
}

/**
 * Thuật toán xếp lịch đấu loại trực tiếp (Knock-out)
 * Tự động tạo cây nhánh đấu. Nếu số đội không phải là lũy thừa của 2 (2, 4, 8, 16...),
 * hệ thống sẽ xếp các đội có thứ hạng cao/may mắn được "Bye" (đi thẳng vào vòng trong).
 */
export function generateKnockout(teams: Team[]): Match[] {
  if (teams.length < 2) return [];

  const list = [...teams];
  const numTeams = list.length;

  // Xác định số lượng vị trí ở vòng 1 (lũy thừa của 2 nhỏ nhất lớn hơn hoặc bằng numTeams)
  let power = 2;
  while (power < numTeams) {
    power *= 2;
  }

  // Số lượng trận đấu "Bye" (đội được đi thẳng)
  const numByes = power - numTeams;
  const matches: Match[] = [];

  // Vòng 1 (Vòng loại đầu tiên)
  // Ví dụ: Nếu có 6 đội, power = 8. Số trận đấu thực sự ở vòng 1 là (6 - (8 - 6)) / 2 = 2 trận đấu.
  // Các đội được đi tiếp (Bye) = 2 đội.
  // Hãy tổ chức các vòng đấu theo cách dễ hiển thị nhất.
  // Định dạng vòng đấu: 
  // - Vòng 1 (Vòng 16, Tứ kết, Bán kết, Chung kết tùy số đội)
  let roundIdx = 1;
  let roundName = "";
  if (power === 2) roundName = "Chung kết";
  else if (power === 4) roundName = "Bán kết";
  else if (power === 8) roundName = "Tứ kết";
  else if (power === 16) roundName = "Vòng 1/8";
  else roundName = `Vòng loại ${power}`;

  // Chuẩn bị các cặp đấu cho Vòng 1
  // Ta sắp xếp các đội xen kẽ. Để đơn giản và khoa học, ta ghép cặp:
  // Đội i vs Đội (numTeams - 1 - i) cho các trận vòng loại, còn lại được Bye.
  const round1Slots: { home: string; away: string; isBye: boolean; winnerPlaceholder?: string }[] = [];
  
  let teamIdx = 0;
  // Ghép các trận đấu thực tế trước
  const numRealMatchesInRound1 = numTeams - power / 2;
  
  for (let m = 0; m < power / 2; m++) {
    if (m < numRealMatchesInRound1) {
      // Trận đấu thực tế giữa 2 đội
      const home = list[teamIdx++];
      const away = list[teamIdx++];
      round1Slots.push({ home: home.id, away: away.id, isBye: false });
    } else {
      // Đội được đi thẳng (Bye)
      const byeTeam = list[teamIdx++];
      round1Slots.push({ home: byeTeam.id, away: "BYE", isBye: true });
    }
  }

  // Tạo các trận đấu vòng 1
  const round1Matches: Match[] = [];
  for (let m = 0; m < round1Slots.length; m++) {
    const slot = round1Slots[m];
    const matchId = `KO-r${roundIdx}-m${m + 1}`;
    
    if (slot.isBye) {
      round1Matches.push({
        id: matchId,
        round: roundIdx,
        roundName,
        homeTeamId: slot.home,
        awayTeamId: "BYE",
        played: true,
        homeScore: 1, // Tự động thắng để đi tiếp
        awayScore: 0,
        homeYellowCards: 0,
        homeRedCards: 0,
        awayYellowCards: 0,
        awayRedCards: 0,
        isBye: true,
      });
    } else {
      round1Matches.push({
        id: matchId,
        round: roundIdx,
        roundName,
        homeTeamId: slot.home,
        awayTeamId: slot.away,
        played: false,
        homeYellowCards: 0,
        homeRedCards: 0,
        awayYellowCards: 0,
        awayRedCards: 0,
      });
    }
  }
  matches.push(...round1Matches);

  // Tạo các vòng đấu tiếp theo cho đến Chung kết
  let currentRoundSlots = power / 2; // Số trận ở vòng vừa rồi
  let prevRoundMatches = round1Matches;

  while (currentRoundSlots > 1) {
    roundIdx++;
    currentRoundSlots /= 2;
    
    let nextRoundName = "";
    if (currentRoundSlots === 1) nextRoundName = "Chung kết";
    else if (currentRoundSlots === 2) nextRoundName = "Bán kết";
    else if (currentRoundSlots === 4) nextRoundName = "Tứ kết";
    else nextRoundName = `Vòng ${currentRoundSlots * 2}`;

    const nextRoundMatches: Match[] = [];
    for (let m = 0; m < currentRoundSlots; m++) {
      const matchId = `KO-r${roundIdx}-m${m + 1}`;
      
      // Xác định nguồn gốc đội thi đấu (Người thắng từ 2 trận đấu vòng trước)
      const prevMatch1Id = prevRoundMatches[m * 2].id;
      const prevMatch2Id = prevRoundMatches[m * 2 + 1].id;

      // Ban đầu khi chưa đá, ta sẽ hiển thị tên đội là placeholder như "Thắng trận X"
      // Chúng ta sẽ lưu thông tin này để xử lý hiển thị hoặc lấy động
      nextRoundMatches.push({
        id: matchId,
        round: roundIdx,
        roundName: nextRoundName,
        homeTeamId: `WIN:${prevMatch1Id}`, // Mã đặc biệt chỉ người thắng trận đấu trước
        awayTeamId: `WIN:${prevMatch2Id}`, // Mã đặc biệt chỉ người thắng trận đấu trước
        played: false,
        homeYellowCards: 0,
        homeRedCards: 0,
        awayYellowCards: 0,
        awayRedCards: 0,
      });
    }
    matches.push(...nextRoundMatches);
    prevRoundMatches = nextRoundMatches;
  }

  return matches;
}

/**
 * Hàm phân giải tên đội bóng thực tế (hỗ trợ cả các slot đặc biệt như WIN:KO-r1-m1 hay BYE)
 */
export function getTeamDisplayName(
  teamId: string,
  teamsList: Team[],
  matchesList: Match[]
): string {
  if (teamId === "BYE" || teamId === "") return "Nghỉ (Bye)";
  
  if (teamId.startsWith("WIN:")) {
    const sourceMatchId = teamId.substring(4);
    const sourceMatch = matchesList.find((m) => m.id === sourceMatchId);
    
    if (!sourceMatch) return "Thắng trận đấu";
    
    // Nếu trận đấu nguồn đã đá, trả về tên đội thắng thực tế
    if (sourceMatch.played) {
      const winnerId = getMatchWinnerId(sourceMatch);
      if (winnerId) {
        const team = teamsList.find((t) => t.id === winnerId);
        return team ? team.name : "Đội thắng";
      }
    }
    
    // Nếu chưa đá, trả về nhãn placeholder gợi ý
    // Ví dụ: "Thắng Trận 1 (Vòng loại)"
    const matchNumber = sourceMatchId.split("-m")[1] || "";
    return `Thắng Trận ${matchNumber} (${sourceMatch.roundName})`;
  }

  const team = teamsList.find((t) => t.id === teamId);
  return team ? team.name : "Không xác định";
}

/**
 * Hàm tìm ID của đội thắng trận đấu
 */
export function getMatchWinnerId(match: Match): string | null {
  if (!match.played) return null;
  
  const hs = match.homeScore ?? 0;
  const as = match.awayScore ?? 0;
  
  if (hs > as) return match.homeTeamId;
  if (as > hs) return match.awayTeamId;
  
  // Trường hợp hòa ở knockout: Phải có hiệp phụ hoặc penalty.
  // Để đơn giản và trực quan, ta mặc định đội chủ nhà thắng,
  // hoặc người dùng tự nhập tỷ số lệch (ví dụ pen 5-4 thì nhập tỷ số 5-4 hoặc bàn thắng tương ứng).
  // Ta mặc định trả về homeTeamId nếu bằng điểm mà là Bye, hoặc homeTeamId làm mặc định.
  return match.homeTeamId;
}

File: scheduler.py
from models import Team, Match
from typing import List, Dict, Optional

def generate_round_robin(teams: List[Team], group_name: Optional[str] = None) -> List[Match]:
    """
    Áp dụng thuật toán xoay vòng (Circle Method) để xếp lịch thi đấu vòng tròn 1 lượt.
    Nếu số lượng đội trong một bảng là lẻ, tự động thêm một đội ảo tên là 'Nghỉ' (BYE).
    """
    if len(teams) < 2:
        return []
        
    team_list = list(teams)
    
    # Thêm đội BYE nếu số lượng đội lẻ
    if len(team_list) % 2 != 0:
        team_list.append(Team(id="BYE", name="Nghỉ", group=group_name))
        
    n = len(team_list)
    num_rounds = n - 1
    matches = []
    
    for r in range(1, num_rounds + 1):
        for i in range(n // 2):
            home_idx = i
            away_idx = n - 1 - i
            home = team_list[home_idx]
            away = team_list[away_idx]
            
            # Luân phiên sân nhà/sân khách cho công bằng
            is_home = (i % 2 == 0) if r % 2 == 1 else (i % 2 == 1)
            first = home if is_home else away
            second = away if is_home else home
            
            match_id = f"{group_name or 'RR'}-R{r}-{first.id}-{second.id}"
            
            if first.id != "BYE" and second.id != "BYE":
                matches.append(Match(
                    id=match_id,
                    round_num=r,
                    round_name=f"Vòng {r}",
                    home_team_id=first.id,
                    away_team_id=second.id,
                    played=False,
                    group=group_name
                ))
            else:
                # Đội thi đấu với BYE sẽ được nghỉ ngơi vòng này
                active_team = second if first.id == "BYE" else first
                matches.append(Match(
                    id=match_id,
                    round_num=r,
                    round_name=f"Vòng {r}",
                    home_team_id=active_team.id,
                    away_team_id="BYE",
                    played=True,
                    home_score=0,
                    away_score=0,
                    group=group_name,
                    is_bye=True
                ))
                
        # Xoay vòng tròn: giữ phần tử đầu tiên cố định, rải đều các phần tử còn lại
        team_list = [team_list[0]] + [team_list[-1]] + team_list[1:-1]
        
    return sorted(matches, key=lambda m: m.round_num)

def generate_knockout(teams: List[Team]) -> List[Match]:
    """
    Xếp cặp đấu loại trực tiếp (Knock-out) theo nhánh cây.
    Xử lý các kích thước đội không phải lũy thừa của 2 bằng cách thêm các slot 'Bye' ở vòng đầu tiên.
    """
    if len(teams) < 2:
        return []
        
    team_list = list(teams)
    num_teams = len(team_list)
    
    # Tính lũy thừa gần nhất của 2 (lớn hơn hoặc bằng số đội)
    power = 2
    while power < num_teams:
        power *= 2
        
    matches = []
    round_idx = 1
    
    # Xác định tên vòng đầu
    if power == 2:
        round_name = "Chung kết"
    elif power == 4:
        round_name = "Bán kết"
    elif power == 8:
        round_name = "Tứ kết"
    elif power == 16:
        round_name = "Vòng 1/8"
    else:
        round_name = f"Vòng loại {power}"
        
    # Chuẩn bị cặp đấu vòng 1
    slots = []
    team_ptr = 0
    num_real_matches = num_teams - power // 2  # Số trận thực sự cần đá ở vòng loại
    
    for m in range(power // 2):
        if m < num_real_matches:
            home = team_list[team_ptr]
            team_ptr += 1
            away = team_list[team_ptr]
            team_ptr += 1
            slots.append({"home": home.id, "away": away.id, "is_bye": False})
        else:
            bye_team = team_list[team_ptr]
            team_ptr += 1
            slots.append({"home": bye_team.id, "away": "BYE", "is_bye": True})
            
    round1_matches = []
    for m, slot in enumerate(slots):
        match_id = f"KO-R{round_idx}-M{m + 1}"
        if slot["is_bye"]:
            # Trận đấu có đội được đặc cách (BYE) đi thẳng
            round1_matches.append(Match(
                id=match_id,
                round_num=round_idx,
                round_name=round_name,
                home_team_id=slot["home"],
                away_team_id="BYE",
                played=True,
                home_score=1,
                away_score=0,
                is_bye=True
            ))
        else:
            round1_matches.append(Match(
                id=match_id,
                round_num=round_idx,
                round_name=round_name,
                home_team_id=slot["home"],
                away_team_id=slot["away"],
                played=False
            ))
    matches.extend(round1_matches)
    
    current_slots = power // 2
    prev_matches = round1_matches
    
    # Xây dựng các vòng tiếp theo cho đến trận Chung kết
    while current_slots > 1:
        round_idx += 1
        current_slots //= 2
        
        if current_slots == 1:
            next_round_name = "Chung kết"
        elif current_slots == 2:
            next_round_name = "Bán kết"
        elif current_slots == 4:
            next_round_name = "Tứ kết"
        else:
            next_round_name = f"Vòng {current_slots * 2}"
            
        next_round_matches = []
        for m in range(current_slots):
            match_id = f"KO-R{round_idx}-M{m + 1}"
            prev_m1_id = prev_matches[m * 2].id
            prev_m2_id = prev_matches[m * 2 + 1].id
            
            # Sử dụng mã WIN:MatchID làm giữ chỗ cho đội thắng của trận đấu trước
            next_round_matches.append(Match(
                id=match_id,
                round_num=round_idx,
                round_name=next_round_name,
                home_team_id=f"WIN:{prev_m1_id}",
                away_team_id=f"WIN:{prev_m2_id}",
                played=False
            ))
        matches.extend(next_round_matches)
        prev_matches = next_round_matches
        
    return matches

def get_team_display_name(team_id: str, teams: List[Team], matches: List[Match]) -> str:
    """Trả về tên hiển thị của đội bóng, giải mã mã WIN:MatchID nếu cần thiết."""
    if team_id == "BYE" or not team_id:
        return "Nghỉ (BYE)"
        
    if team_id.startswith("WIN:"):
        source_id = team_id[4:]
        source_match = next((m for m in matches if m.id == source_id), None)
        if not source_match:
            return "Thắng trận đấu"
            
        if source_match.played:
            winner_id = get_match_winner_id(source_match)
            if winner_id:
                winner_team = next((t for t in teams if t.id == winner_id), None)
                if winner_team:
                    return winner_team.name
            return "Đội thắng"
            
        # Nếu chưa đá, trả về nhãn giữ chỗ
        match_num = source_id.split("-M")[-1] if "-M" in source_id else ""
        return f"Thắng Trận {match_num} ({source_match.round_name})"
        
    team = next((t for t in teams if t.id == team_id), None)
    return team.name if team else "Chưa xác định"

def get_match_winner_id(match: Match) -> Optional[str]:
    """Tìm ID của đội thắng trận đấu."""
    if not match.played:
        return None
    hs = match.home_score if match.home_score is not None else 0
    as_ = match.away_score if match.away_score is not None else 0
    if hs > as_:
        return match.home_team_id
    elif as_ > hs:
        return match.away_team_id
    return match.home_team_id  # Mặc định chủ nhà nếu có hòa
